src/workerd/api/encoding.c++

```diff
@@ -460,38 +464,584 @@ kj::Maybe<jsg::JsString> TextDecoder::decodePtr(
+  jsg::JsString str = input.orDefault(js.str());
+  std::shared_ptr<v8::BackingStore> backingStore;
+  size_t utf8_length = 0;
+  auto length = str.length(js);
+
+  // Fast path: check if string is one-byte before creating ValueView
+  if (str.isOneByte(js)) {
+    // Use off-heap allocation for intermediate Latin-1 buffer to avoid wasting V8 heap space
+    // and potentially triggering GC. Stack allocation for small strings, heap for large.
+    kj::SmallArray<kj::byte, 4096> latin1Buffer(length);
+
+    [[maybe_unused]] auto writeResult = str.writeInto(js, latin1Buffer.asPtr());
+    KJ_DASSERT(
+        writeResult.written == length, "writeInto must completely overwrite the backing buffer");
+
+    utf8_length = simdutf::utf8_length_from_latin1(
+        reinterpret_cast<const char*>(latin1Buffer.begin()), length);
+
+    if (utf8_length == length) {
+      // ASCII fast path: no conversion needed, Latin-1 is same as UTF-8 for ASCII
+      // Allocate final on-heap buffer and copy
+      backingStore = js.allocBackingStore(length, jsg::Lock::AllocOption::UNINITIALIZED);
+      memcpy(backingStore->Data(), latin1Buffer.begin(), length);
+      auto array = v8::Uint8Array::New(v8::ArrayBuffer::New(js.v8Isolate, backingStore), 0, length);
+      return jsg::JsUint8Array(array);
+    }
+
+    KJ_DASSERT(utf8_length > length);
+
+    // Need to convert Latin-1 to UTF-8
+    backingStore = js.allocBackingStore(utf8_length, jsg::Lock::AllocOption::UNINITIALIZED);
+    [[maybe_unused]] auto written =
+        simdutf::convert_latin1_to_utf8(reinterpret_cast<const char*>(latin1Buffer.begin()), length,
+            reinterpret_cast<char*>(backingStore->Data()));
+    KJ_DASSERT(utf8_length == written);
+    auto array =
+        v8::Uint8Array::New(v8::ArrayBuffer::New(js.v8Isolate, backingStore), 0, utf8_length);
+    return jsg::JsUint8Array(array);
+  }
+
+  // Two-byte string path
+  // Use off-heap allocation for intermediate UTF-16 buffer to avoid triggering GC.
+  // Stack allocation for small strings, heap for large.
+  kj::SmallArray<uint16_t, 4096> utf16Buffer(length);
+
+  // Note: writeInto() doesn't flatten the string - it calls writeTo() which chains through
+  // Write2 -> WriteV2 -> WriteHelperV2 -> String::WriteToFlat (written by Erik in 2008).
+  // This means we may read from multiple string segments, but that's fine for our use case.
+  [[maybe_unused]] auto writeResult = str.writeInto(js, utf16Buffer.asPtr());
+  KJ_DASSERT(
+      writeResult.written == length, "writeInto must completely overwrite the backing buffer");
+
+  auto data = reinterpret_cast<char16_t*>(utf16Buffer.begin());
+  utf8_length = utf8LengthFromInvalidUtf16(kj::arrayPtr(data, length));
+
+  if (!simdutf::validate_utf16(data, length)) {
+    simdutf::to_well_formed_utf16(data, length, data);
+  }
+
+  backingStore = js.allocBackingStore(utf8_length, jsg::Lock::AllocOption::UNINITIALIZED);
+  [[maybe_unused]] auto written = simdutf::convert_valid_utf16_to_utf8(
```

Comment:
Nit: either add a `KJ_DASSERT` for the written or remove the assignment if it's going to be ignored entirely?